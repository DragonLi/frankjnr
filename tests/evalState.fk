interface State X = get : X
	          | put : X -> Unit

evalState : X -> <State X>Y    -> Y
evalState   x    <put x' -> k>  = evalState x' (k unit)
evalState   x    <get    -> k>  = evalState x (k x)
evalState   x         y         = y

main : String
main! = evalState "Hello" (put (strcat get! " World!\n"); get!)
