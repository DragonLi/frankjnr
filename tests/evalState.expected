MkRawProg
 [MkRawItfTm (MkItf "State" ["x"]
           	   [MkCmd "get" (MkCType [] (MkPeg MkOpenAb (MkTVar "x")))
 	  	   ,MkCmd "put" (MkCType [MkPort MkIdAdj (MkTVar "x")]
		   	      	         (MkPeg MkOpenAb (MkTVar "Unit")))])
 ,MkRawSigTm (MkSig "evalState"
   (MkCType [ MkPort MkIdAdj (MkTVar "X")
            , MkPort (MkAdjPlus MkIdAdj "State" [MkTVar "X"]) (MkTVar "Y")]
            (MkPeg MkOpenAb (MkTVar "Y"))))
 ,MkRawDefTm (MkMHCls "evalState"
 	     	      (MkRawCls [MkVPat (MkVarPat "x")
                                ,MkCmdPat "put" [MkVarPat "x'"] "k"]
 	     (MkRawComb "evalState" [MkRawId "x'"
	     	    	            ,MkRawComb "k" [MkRawId "Unit"]])))
 ,MkRawDefTm (MkMHCls "evalState"
 	     	      (MkRawCls [MkVPat (MkVarPat "x")
                                ,MkCmdPat "get" [] "k"]
 	     (MkRawComb "evalState" [MkRawId "x"
	     	    	            ,MkRawComb "k" [MkRawId "x"]])))
 ,MkRawDefTm (MkMHCls "evalState" (MkRawCls [MkVPat (MkVarPat "x")
                                            ,MkVPat (MkVarPat "y")]
					    (MkRawId "y")))
 ,MkRawSigTm (MkSig "main" (MkCType [] (MkPeg MkOpenAb MkStringTy)))
 ,MkRawDefTm (MkMHCls "main" (MkRawCls []
 	     	      	       (MkRawComb "evalState"
			         [MkRawStr "Hello"
			       	 ,MkRawTmSeq (MkRawComb "put"
				 	       [MkRawComb "strcat"
                                                 [MkRawComb "get" []
					         ,MkRawStr " World!\n"]])
				 	     (MkRawComb "get" [])])))]
